WPMedia/fusion-news-theme-blocks
# Themes 1289 V2 Video Center | Arc block - video needs to resize to handle the correct size when we have vertical video

**This PR goes with an `arc-themes-components` PR. You can find it [here](https://github.com/WPMedia/arc-themes-components/pull/201).**

## Description

Themes V2 has an issue with resizing vertical videos. The tall aspect ratio caused vertical videos to occupy the correct width on the screen, but they were way too tall. This impacted the Video Center, Lead Art, and Article Body Blocks.  
  
After investigating the issue, I found that the current resizing method/styling in the `Video` component is correct. However, I noticed that the blocks that had issues only passed in 16:9 aspect ratios (they were hardcoded to do this instead of passing in the correct aspect ratio of the video). This made vertical videos occupy the width of their, but they got really tall because of the resizing strategy.  
  
My goal for this ticket is to make it so that the affected blocks pass in correct `aspectRatio` values for the videos they are displaying. To do that, I made the following changes:

* In the `default.jsx` files for the `article-body-block`, `lead-art-block`, and `video-player-block` directories:
    * ~~Created a helper function called `gcd` to find the Greatest Common Denominator (GCD) of two numbers~~ **Update:** `gcd` (which is a helper for the `getAspectRatio` function) has been moved into a `utils` directory in `arc-themes-components`. This reduces some code duplication that existed before. The PR can be found [here](https://github.com/WPMedia/arc-themes-components/pull/201)
        * *Note* that because of the "lexical declaration in a case block" error set to throw for switch statements, the default aspect ratio (16:9) is declared above the switch statement in `article-body-block/chains/article-body/default.jsx` (and not close to where the aspect ratio is actually generated)
    * Set 16:9 as a fallback value for aspect ratio. This happens when a promo item cannot be found or when a video has height `0`
    * If the video exists and has existing promo items, the aspect ratio is calculated using the height and width of the video, along with `gcd` to reduce the fraction as much as possible
    * `aspectRatio` is still passed into `Video` as a string with the format `/\d+:\d+/` (i.e., "16:9", "4:3", and "3:2")
* Added tests in `default.tsx` for the `article-body-block`, `lead-art-block`, and `video-player-block` directories:
    * This was necessary because the added logic dropped the test coverage below 85%
    * The Video Player block has a test that makes sure the HTML embed for a vertical video remains the same after a `Video` component is created
    * All three blocks have a test to verify that `aspectRatio` is calculated correctly for a video (each block has a different set of video dimensions to test different cases)
    * `getAspectRatio` is also tested thoroughly in `arc-themes-components/src/utils/get-aspect-ratio/index.test.js`
    * *This is my first time writing tests in React and for Arc, so if you have any feedback please let me know*

## Jira Ticket

[THEMES-1289](https://arcpublishing.atlassian.net/browse/THEMES-1289)

## Acceptance Criteria

1. When we upload a vertical video, the block size should not expand the video, and the video should show vertical black bars should be shown as needed.

## Test Steps

1. Checkout this branch: `git checkout THEMES-1289-vertical-video-resizing`
2. Run fusion repo with linked blocks `npx fusion start -f -l "@wpmedia/video-player-block","@wpmedia/lead-art-block","@wpmedia/article-body-block"`
3. Open your local PageBuilder instance and create a new page.
    * The layout doesn't matter (if you want to copy me exactly, I used the Right Rail layout and put all of the blocks in the "fullwidth2" section)
    * Set the global content source to `content-api` and set `_id` to "JLIWHMWKYJBNBGNYQDMAUXYGOA". This story has embedded videos and lead art for testing
    * Place a Lead Art block
    * Place an Article Body chain. I set it to lazy load
    * Place a Video Center Player block. Under Configure Content, set the Video Content to `ans-item`, Content Source to `content-api`, and `_id` to "3509db8e-aaea-4264-8efa-b9fe039af027". **Make sure the "Inherit gobal content" checkbox is unchecked!** Under display settings, make the title "Rain on the Road", description "Fall", alert badge "rain", and the video display style "Inline Video." This block will be used to test vertical video.
    * Place another Video Center Player block. Under Configure Content, set the Video Content to `ans-item`, Content Source to `content-api`, and `_id` to "4ae0b7de-dc1c-440a-9655-08eb5bfc81fd". **Make sure the "Inherit gobal content" checkbox is unchecked!** Under display settings, make the title "Horizontal Video", description "Real Madrid flag", alert badge "soccer", and the video display style "Inline Video." This block will be used to verify that horizontal videos weren't broken by my changes.
4. Verify that all vertical and horizontal videos in the test page are sized correctly and have regular aspect ratios
5. Open the "All Blocks-Test-TA-Sandbox" page. Edit the "Image Ratio" setting under the Extra Large Promo and Large Promo (the headline of the article is "Explorers Discover", I chose these promo blocks because they currently have videos embedded in them), and make sure that it still works as expected

## Effect Of Changes

### Before

https://github.com/WPMedia/arc-themes-blocks/assets/103296586/522674e8-70df-4c1a-b794-7bf7210caf4a

### After

https://github.com/WPMedia/arc-themes-blocks/assets/103296586/ab0a2294-48e8-4078-990f-d6a3883d265c

## Dependencies or Side Effects

There is an `arc-themes-components` PR. The `video` component's `name` prop can now be any string, not an enum (`["16:9", "3:2", "4:3"]`). There is also a new `get-aspect-ratio` utility function that calculates the aspect ratio of something given its' width and height. The utility function is tested too. You can find the PR [here](https://github.com/WPMedia/arc-themes-components/pull/201).

## Author Checklist

        please explain why (so that we can fix it whenever it gets refactored). **N/A**

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._



[THEMES-1289]: https://arcpublishing.atlassian.net/browse/THEMES-1289?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
